[tests] Refactor PostgreSql tests to use testcontainers

radical opened the pull request:
And use `ContainerImageTags.PostgreSql` instead of hardcoding the `image:tag`.
And move ContainerImageTags from Aspire.Components.Common to Aspire.Hosting

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2862)

radical committed:
Refactor PostgreSQL tests to use testcontainers

radical committed:
Use ContainerImageTags.PostgreSql instead of hardcoding the image:tag

# Conflicts:
#	src/Aspire.Hosting/ContainerImageTags.cs

radical commented:
Depends on https://github.com/dotnet/aspire/pull/2856 .

radical committed:
Merge main

radical committed:
Move ContainerImageTags from Aspire.Components.Common to Aspire.Hosting

radical committed:
Remove unncessary RequiresDocker attributes

.. and use a fallback localhost connection string when docker isn't
available.

radical committed:
cleanup

eerhardt commented on tests/Aspire.Npgsql.Tests/ConformanceTests.cs:
```diff
@@ -35,7 +35,8 @@ public class ConformanceTests : ConformanceTests<NpgsqlDataSource, NpgsqlSetting
     protected override bool SupportsKeyedRegistrations => true;
 
-    protected override bool CanConnectToServer => s_canConnectToServer.Value;
+    protected override bool CanCreateClientWithoutConnectingToServer => false;
```
Why do we need this change?

radical replied:
This property defaults to `true`. Due to that:
https://github.com/dotnet/aspire/blob/16679d35f9ab64d7221033888a94ec495780aead/tests/Aspire.Components.Common.Tests/ConformanceTests.cs#L464-L470

.. if you have `ConnectToServer=false` - no docker available - then it wouldn't skip the test and instead fail for some tests. Now that I think about it, this should be used in tests that don't depend on really connecting to the server. So, either it is incorrectly used in some test or I missed something.

I ran the tests again, and they pass even if this is `true`. In fact, we have this for redis also but isn't needed. Looking at it, IIUC I blindly copied this bit from your rabbitmq PR, which does need this setting!

eerhardt commented on tests/Aspire.Npgsql.Tests/ConformanceTests.cs:
```diff
@@ -116,54 +120,25 @@ public void BothDataSourceAndConnectionCanBeResolved(string? key)
-    {
-        SkipIfCanNotConnectToServer();
-
-        RemoteExecutor.Invoke(() => ActivitySourceTest(key: "key")).Dispose();
```
Why don't we need the RemoteExecutor anymore to run these tests?

eerhardt commented on tests/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests.csproj:
```diff
@@ -6,10 +6,14 @@
   <ItemGroup>
+    <Compile Include="$(RepoRoot)src\Aspire.Hosting\ContainerImageTags.cs" />
+    <Compile Include="..\Aspire.Npgsql.Tests\PostgreSQLContainerFixture.cs" />
     <None Include="$(RepoRoot)src\Components\Aspire.Npgsql.EntityFrameworkCore.PostgreSQL\ConfigurationSchema.json" CopyToOutputDirectory="PreserveNewest" />
```
(nit) let's be consistent on the order of the json file and the Compile items. These are in a different order than Aspire.Npgsql.Tests.

eerhardt commented on tests/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/ConformanceTests.cs:
```diff
@@ -3,23 +3,22 @@
-    protected const string ConnectionString = "Host=localhost;Database=test;Username=postgres;Password=postgres";
-
-    private static readonly Lazy<bool> s_canConnectToServer = new(GetCanConnect);
+    protected readonly PostgreSQLContainerFixture _containerFixture;
```
```suggestion
    private readonly PostgreSQLContainerFixture _containerFixture;
```
? I don't see anyone using this.

eerhardt commented on tests/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/ConformanceTests.cs:
```diff
@@ -126,28 +129,21 @@ public void DbContextCanBeAlwaysResolved()
+    public Task TracingEnablesTheRightActivitySource()
+        => RunWithFixtureAsync(obj => obj.ActivitySourceTest(key: null));
 
-        RemoteExecutor.Invoke(() => ActivitySourceTest(key: null)).Dispose();
```
Why doesn't this test need RemoteExecutor anymore?

radical replied:
IIUC, we were running these tests with RemoteExecutor to run in a separate process so changes to the db don't affect other tests. And that wasn't working because RemoteExecutor would complain about not being able to marshal `_containerFixture` field. So, I create a fresh container and run the test with that.

 But now reading the comments:
```
    // This method can have side effects (setting AppContext switch, enabling activity source by name).
    // That is why it needs to be executed in a standalone process.
    // We use RemoteExecutor for that, but it does not support abstract classes
    // (it can not determine the type to instantiate), so that is why this "test"
    // is here and derived types call it
    protected void ActivitySourceTest(string? key)
```

.. the above suggests that this would affect the running process anyway, so `RemoteExecutor` should be used. I'll fix it.

radical committed:
Update tests/Aspire.Npgsql.EntityFrameworkCore.PostgreSQL.Tests/ConformanceTests.cs

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Address review feedback from @ eerhardt

radical committed:
Address review feedback from @ eerhardt

radical committed:
Address review feedback from @ eerhardt

Bring back `RemoteExecutor` because:

```
    // This method can have side effects (setting AppContext switch, enabling activity source by name).
    // That is why it needs to be executed in a standalone process.
    // We use RemoteExecutor for that, but it does not support abstract classes
    // (it can not determine the type to instantiate), so that is why this "test"
    // is here and derived types call it
    protected void ActivitySourceTest(string? key)
```

radical committed:
Merge remote-tracking branch 'origin/main' into postgres-testcontainers

radical committed:
Merge remote-tracking branch 'origin/main' into postgres-testcontainers

radical committed:
Merge remote-tracking branch 'origin/main' into postgres-testcontainers

radical committed:
Track file path change

radical committed:
Merge remote-tracking branch 'origin/main' into postgres-testcontainers

# Conflicts:
#	src/Aspire.Hosting.AppHost/ContainerImageTags.cs
#	tests/Aspire.Hosting.Tests/ContainerResourceBuilderTests.cs
#	tests/Aspire.StackExchange.Redis.Tests/RedisContainerFixture.cs

radical committed:
Track changes from main to ContainerImageTags

radical committed:
Dispose RemoteExecutor

eerhardt commented on tests/Aspire.Npgsql.Tests/ConformanceTests.cs:
```diff
@@ -116,54 +120,25 @@ public void BothDataSourceAndConnectionCanBeResolved(string? key)
-        NpgsqlConnection connection = new(ConnectionSting);
-        NpgsqlCommand? cmd = null;
-
+        var fixture = new PostgreSQLContainerFixture();
+        await fixture.InitializeAsync();
         try
```
You can get rid of the `try-finally` if you use

```C#
await using var fixture = new PostgreSQLContainerFixture();
await fixture.InitializeAsync();
test(new ConformanceTests(fixture));
```

radical committed:
Address review feedback from @ eerhardt

radical committed:
Merge branch 'dotnet:main' into postgres-testcontainers

radical committed:
Merge remote-tracking branch 'origin/main' into postgres-testcontainers

radical committed:
Use remoteExecutor for tracing tests

eerhardt reviewed (approved):
LGTM
